### PR TITLE
support data_norm gpu kernel

### PR DESCRIPTION
**Pull request by hutuxian**
Previously, DataNorm OP was only able to run in CPU. This PR implements its GPU kernel.
What's more, in GPU Kernel, we update three summary messages in its grad part.

### CODE REVIEW

File: paddle/fluid/operators/data_norm_op.cc [hunk 1]
```diff
@@ -129,7 +129,14 @@ class DataNormOpMaker : public framework::OpProtoAndCheckerMaker {
                  "(int, default -1) Dimension of one slot if set, "
                  "when the input is concated by slot-wise embeddings")
         .SetDefault(-1);
+    AddAttr<float>(
+        "summary_decay_rate",
+        "(float, default 0.9999999) The decay rate when update the summary")
+        .SetDefault(0.9999999);
     AddAttr<std::string>("data_layout", "").SetDefault("NCHW");
+    AddAttr<bool>("sync_with_nccl",
```
Comment (gavin1332) [inline paddle/fluid/operators/data_norm_op.cc]: would `sync_stats` be better? nccl is one method but not the only one.

File: paddle/fluid/operators/data_norm_op.cu [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+#define CUDA_KERNEL_LOOP(i, n)                                 \
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < (n); \
+       i += blockDim.x * gridDim.x)
+
+inline int GET_BLOCKS(const int N) {
+  return (N + PADDLE_CUDA_NUM_THREADS - 1) / PADDLE_CUDA_NUM_THREADS;
+}
+
+template <typename T>
+__global__ void kernel_data_norm_ff(int N, int C, const T *x, T *y,
```
Comment (gavin1332) [inline paddle/fluid/operators/data_norm_op.cu]: illegal function naming. According to the naming conventions of paddle, KernelDataNormFF is preferred

File: paddle/fluid/operators/data_norm_op.cu [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                                    const T *mean, const T *scale) {
+  CUDA_KERNEL_LOOP(i, N * C) {
+    int col = i % C;
+    y[i] = (x[i] - mean[col]) * scale[col];
+  }
+}
+
+template <typename T>
+__global__ void kernel_mean_scale(int C, const T *batch_size,
```
Comment (gavin1332) [inline paddle/fluid/operators/data_norm_op.cu]: KernelMeanScale

File: paddle/fluid/operators/data_norm_op.cu [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+                                  const T *batch_sum, const T *batch_square_sum,
+                                  T *mean, T *scale) {
+  CUDA_KERNEL_LOOP(i, C) {
+    mean[i] = batch_sum[i] / batch_size[i];
+    scale[i] = sqrt(batch_size[i] / batch_square_sum[i]);
+  }
+}
+
+template <typename T>
+__global__ void kernel_data_norm_bp(int N, int C, const T *y_grad,
```
Comment (gavin1332) [inline paddle/fluid/operators/data_norm_op.cu]: KernelDataNormBP

File: paddle/fluid/operators/data_norm_op.cu [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+                                    const T *scale, T *x_grad) {
+  CUDA_KERNEL_LOOP(i, N * C) { x_grad[i] = y_grad[i] * scale[i % C]; }
+}
+
+template <typename T>
+__global__ void kernel_data_norm_bp_stat(int N, int C, const T *x_val,
```
Comment (gavin1332) [inline paddle/fluid/operators/data_norm_op.cu]: KernelDataNormBPStat

File: paddle/fluid/operators/data_norm_op.cu [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+          (x_val[j * C + i] - means[i]) * (x_val[j * C + i] - means[i]);
+    }
+    batch_size[i] = 1;
+    batch_sum[i] = val_sum / N;
+    batch_square_sum[i] = square_sum / N + squared_sum_epsilon;
+  }
+}
+
+template <typename T>
+__global__ void kernel_update_param(int C, const T *d_batch_size,
```
Comment (gavin1332) [inline paddle/fluid/operators/data_norm_op.cu]: KernelUpdataParam

File: paddle/fluid/operators/data_norm_op.cu [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+          reinterpret_cast<const void *>(d_batch_sum),
+          reinterpret_cast<void *>(d_batch_sum), C,
+          platform::ToNCCLDataType(x->type()), ncclSum, comm->comm(),
+          comm_stream));
+      PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+          reinterpret_cast<const void *>(d_batch_square_sum),
+          reinterpret_cast<void *>(d_batch_square_sum), C,
+          platform::ToNCCLDataType(x->type()), ncclSum, comm->comm(),
+          comm_stream));
+      e_sync = cudaStreamSynchronize(comm_stream);
```
Comment (gavin1332) [inline paddle/fluid/operators/data_norm_op.cu]: as synchronization is required, you would better to use calculation stream without call synchronizing.

**Review by gavin1332 — COMMENTED** [review]
Missing unittest to test sync_with_nccl?

File: paddle/fluid/operators/data_norm_op.cu (hunk 7 above)
Comment (hutuxian) [reply paddle/fluid/operators/data_norm_op.cu]: I think so, thanks

**Review by gavin1332 — APPROVED** [review]
LGTM

**Comment by gavin1332** [comment]
oh, it seems you forget to update the document since you have changed the API

**Comment by hutuxian** [comment]
> oh, it seems you forget to update the document since you have changed the API

Will post the Chinese document here.

**Review by Boyan-Liu — COMMENTED** [review]
Don't forget to change the Chinese document.

**Review by Superjomn — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]
LGTM

**Comment by hutuxian** [comment]
image

**Comment by hutuxian** [comment]
image

**Review by Boyan-Liu — APPROVED** [review]
LGTM

**Comment by hutuxian** [comment]
Chinese Document： https://github.com/PaddlePaddle/FluidDoc/pull/1618

**Pull request merged**